### DISCUSSION TITLE
Duplicate logs

### DISCUSSION THREAD

**Discussion by chris-fibo** [sqlalchemy/sqlalchemy#8374] [Usage Questions]
With sqlalchemy I am getting duplicate logs. One goes to stdout and one goes to my logs.
But I can't seem to find a way to turn off the logging to stdout.
If I set echo=False I get no log messages.

It seems like, no matter what the setting is InstanceLogger will ALWAYS add a default handler.
The __init__ has the following code:
`        # if echo flag is enabled and no handlers,
        # add a handler to the list
        if self._echo_map[echo] <= logging.INFO and not self.logger.handlers:
            _add_default_handler(self.logger)
`
The comment says if the echo flag is true. Except that self._echo_map[echo] <= logging.INFO will always be true. And the first time this is called, self.logger.handlers will be empty. Therefore a default handler will ALWAYS be added.

I think that is a separate issue. I  think I want echo to be true, but no default handler to be added.

**Comment by CaselIT** [comment]
Hi,

Can't say I can reproduce the issue. Can you provide an example script that illustrates the issue?

**Reply by chris-fibo** [reply]
I'll see if I can't do something minimal that repros this.
But for the moment, I am running my own logger that logs to stdout. It production that logger goes elsewhere and we don't even see stdout. But in debug mode I just dump the log to stdout. But with echo set to True or 'debug' SQLAlchemy adds its own handler, and I get two logs. If I set echo to None or False I get no logs from SQLAlchemy.
Right now I'm "fixing" it, by simply having SQLAlchemy not create the logging handler.

**Comment by zzzeek** [comment]
try configuring logging [without using the echo flag](https://docs.sqlalchemy.org/en/14/core/engines.html#configuring-logging).  it sounds like your application already has some logging configuration in place that's getting involved, so just add SQLAlchemy logging to that without using "echo" at all.

**Reply by chris-fibo** [reply]
If I set echo to None or False, I don't get any logging from SQLAlchemy. 
Sorry, I should have been more explicit. Yes, I have my own logging. And in debug mode that logging goes to stdout. I've also got a nice formatter for it.
But then if I have echo turned on (either True or 'debug') SQLAlchemy sends it both to my logger as well as its own logger.
If I set echo to None or False, then I don't get any logging for SQLAlchemy.

**Reply by zzzeek** [reply]
basically dont turn echo on. it's not supported when you have other logging in place already.

**Reply by CaselIT** [reply]
if you want to customer sqlalchemy level using an external configured logger you can just do
`logging.getLogger('sqlalchemy.engine').setLevel(logging.INFO)`

**Reply by chris-fibo** [reply]
Ahhh thanks. I need to set echo=False/None AND then set the level for sqlaclhemy.engine to INFO.
